File: Python3/0056_merge_intervals.py
# https://leetcode-cn.com/problems/merge-intervals/

class Solution:
    def merge(self, intervals):
        """
        Inputs:
            intervals: list[list[int]]
        Outputs:
            res: list[list[int]]
        """

        # Sort the intervals based on their first elements.
        intervals.sort()

        # Add the first interval into result.
        res = [intervals[0]]
        idx = 0
        right = intervals[0][1]

        for i in range(1, len(intervals)):
            # Merge the two intervals.
            if right >= intervals[i][0]:
                res[idx][1] = max(right, intervals[i][1])
            
            # Add the interval into result.
            else:
                res.append(intervals[i])
                idx = idx + 1

            right = res[idx][1]

        return res

File: Cpp/0056_merge_intervals.cpp
// https://leetcode.cn/problems/merge-intervals/

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(),
            [](const vector<int>& interval1, const vector<int>& interval2) {
                return interval1[0] < interval2[0];
            });
        
        vector<vector<int>> res = {intervals[0]};
        int right = intervals[0][1];
        for (int i = 1; i < intervals.size(); ++i) {
            if (right < intervals[i][0]) {
                res.push_back(intervals[i]);
                right = intervals[i][1];
            }
            else if (right < intervals[i][1]) {
                right = intervals[i][1];
                res.back()[1] = right;
            }
        }
        return res;
    }
};
